# ML-04 — Search Intelligence Data Contract

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/haneen06-blip/flyrank-assiment/blob/main/work/notebooks/w03_data_contract.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [34]:
%pip install -q duckdb huggingface_hub pandas scikit-learn matplotlib

In [35]:
from huggingface_hub import login
from google.colab import userdata

HF_TOKEN = userdata.get("HF_TOKEN")

login(token=HF_TOKEN)

print("Hugging Face login successful.")

Hugging Face login successful.


In [36]:
import duckdb

con = duckdb.connect()

print("DuckDB connected.")

DuckDB connected.


In [37]:
tables = con.execute("SHOW TABLES").fetchdf()
tables

,name


In [38]:
# Check what variables are already available
%whos

Variable                 Type                      Data/Info
------------------------------------------------------------
DecisionTreeClassifier   ABCMeta                   <class 'sklearn.tree._cla<...>.DecisionTreeClassifier'>
HF_TOKEN                 str                       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
REL                      str                       hf://datasets/FlyRank/internship-warehouse
TABLES                   dict                      n=5
X_honest                 DataFrame                         gsc_impressions  <...>[331436 rows x 5 columns]
X_leaky                  DataFrame                         gsc_impressions  <...>[331436 rows x 6 columns]
X_test_h                 DataFrame                         gsc_impressions  <...>n[66288 rows x 5 columns]
X_test_l                 DataFrame                         gsc_impressions  <...>n[66288 rows x 6 columns]
X_train_h                DataFrame                         gsc_impressions  <...>[265148 rows x 5 columns

In [39]:
import duckdb

con = duckdb.connect()

con.execute(
    f"CREATE OR REPLACE SECRET hf "
    f"(TYPE huggingface, TOKEN '{HF_TOKEN}')"
)

REL = "hf://datasets/FlyRank/internship-warehouse"

TABLES = {
    "dim_clients": f"read_parquet('{REL}/dim_clients.parquet')",
    "dim_content": f"read_parquet('{REL}/dim_content.parquet')",
    "fact_daily": f"read_parquet('{REL}/fact_content_daily_performance/**/*.parquet')",
    "fact_daily_sample": f"read_parquet('{REL}/fact_content_daily_performance_sample.parquet')",
    "fact_query_90d": f"read_parquet('{REL}/fact_content_query_90d.parquet')",
}

for name, src in TABLES.items():
    n = con.sql(f"SELECT COUNT(*) FROM {src}").fetchone()[0]
    print(f"{name:22} {n:>12,} rows")

dim_clients                     104 rows
dim_content                 519,606 rows


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

fact_daily               78,835,655 rows
fact_daily_sample        11,694,072 rows
fact_query_90d            2,414,248 rows


In [40]:
con.sql("""
DESCRIBE SELECT *
FROM read_parquet(
    'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/**/*.parquet'
)
""").fetchdf()

,column_name,column_type,null,key,default,extra
0,report_date,DATE,YES,None,None,None
1,client_hash_id,VARCHAR,YES,None,None,None
2,content_hash_id,VARCHAR,YES,None,None,None
3,client_has_gsc,BOOLEAN,YES,None,None,None
4,client_has_ga4,BOOLEAN,YES,None,None,None
5,gsc_data_available,BOOLEAN,YES,None,None,None
6,ga4_data_available,BOOLEAN,YES,None,None,None
7,gsc_impressions,BIGINT,YES,None,None,None
8,gsc_clicks,BIGINT,YES,None,None,None
9,gsc_sum_position,BIGINT,YES,None,None,None


## 1. Unit of analysis + time window

One row represents one content item for one client on one report date. For this assignment, I use the March 2026 window (month = '2026-03') as the mid-panel month.

In [41]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
# Query 1 — Verify the observed grain for March 2026

grain_check = con.sql("""
SELECT
    COUNT(*) AS rows,
    COUNT(DISTINCT
        CONCAT(
            CAST(report_date AS VARCHAR), '|',
            client_hash_id, '|',
            content_hash_id
        )
    ) AS distinct_grains
FROM read_parquet(
    'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/**/*.parquet'
)
WHERE month = '2026-03'
""").fetchdf()

grain_check


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,rows,distinct_grains
0,9841378,9841378


## 2. Fields: feature / label / context / excluded

**Features:**
- gsc_impressions
- gsc_clicks
- gsc_avg_position
- ga4_sessions
- ga4_engaged_sessions

**Label / proxy:**
A future decline proxy: a content item is marked as declining when its April 2026 GSC impressions are lower than its March 2026 GSC impressions.

**Context:**
- report_date
- client_hash_id
- content_hash_id
- client_has_gsc
- client_has_ga4
- gsc_data_available
- ga4_data_available
- month

**Excluded:**
Future-period performance fields and any future-derived field used to construct the outcome. They are excluded from the final features because they would not be known at the March decision moment and would cause target leakage.

In [42]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
decision_frame = con.sql("""
WITH march AS (
    SELECT
        client_hash_id,
        content_hash_id,
        SUM(gsc_impressions) AS gsc_impressions,
        SUM(gsc_clicks) AS gsc_clicks,
        AVG(gsc_avg_position) AS gsc_avg_position,
        SUM(ga4_sessions) AS ga4_sessions,
        SUM(ga4_engaged_sessions) AS ga4_engaged_sessions
    FROM read_parquet(
        'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/**/*.parquet'
    )
    WHERE month = '2026-03'
    GROUP BY client_hash_id, content_hash_id
),
april AS (
    SELECT
        client_hash_id,
        content_hash_id,
        SUM(gsc_impressions) AS april_impressions
    FROM read_parquet(
        'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/**/*.parquet'
    )
    WHERE month = '2026-04'
    GROUP BY client_hash_id, content_hash_id
)
SELECT
    m.*,
    a.april_impressions,
    CASE
        WHEN a.april_impressions < m.gsc_impressions THEN TRUE
        ELSE FALSE
    END AS declining_proxy
FROM march m
LEFT JOIN april a
    USING (client_hash_id, content_hash_id)
""").fetchdf()

decision_frame.head()


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,client_hash_id,content_hash_id,gsc_impressions,gsc_clicks,gsc_avg_position,ga4_sessions,ga4_engaged_sessions,april_impressions,declining_proxy
0,client_73cda7b4e4f265ea,content_b7e512995f79d5a6,1140.0,2.0,4.394234,0.0,0.0,1151.0,False
1,client_73cda7b4e4f265ea,content_05597932fe4da067,57.0,0.0,2.714744,0.0,0.0,73.0,False
2,client_73cda7b4e4f265ea,content_905aa32a0230694e,149.0,0.0,6.481453,4.0,0.0,98.0,True
3,client_73cda7b4e4f265ea,content_05434271b257bb68,1421.0,6.0,6.320337,9.0,0.0,2275.0,False
4,client_73cda7b4e4f265ea,content_d056587ff7faca0c,2770.0,16.0,4.459107,3.0,0.0,6266.0,False


### Five features and availability

1. **gsc_impressions** — Available at the March decision moment because it comes from March GSC observations.
2. **gsc_clicks** — Available at the March decision moment because it comes from March GSC observations.
3. **gsc_avg_position** — Available at the March decision moment because it comes from March GSC observations.
4. **ga4_sessions** — Available at the March decision moment because it comes from March GA4 observations.
5. **ga4_engaged_sessions** — Available at the March decision moment because it comes from March GA4 observations.

## 3. Verify it with queries (grain, counts, missing values, windows)

I use March 2026 as the mid-panel decision month. The checks below verify the observed grain, the March row count and date span, and data availability using IS TRUE.

In [43]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
# Query 1 — Grain
grain_check = con.sql("""
SELECT
    COUNT(*) AS rows,
    COUNT(DISTINCT
        CONCAT(
            CAST(report_date AS VARCHAR), '|',
            client_hash_id, '|',
            content_hash_id
        )
    ) AS distinct_grains
FROM read_parquet(
    'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/**/*.parquet'
)
WHERE month = '2026-03'
""").fetchdf()

print("Query 1 — Grain")
grain_check


# Query 2 — March row count and date span
march_window = con.sql("""
SELECT
    COUNT(*) AS row_count,
    MIN(report_date) AS min_report_date,
    MAX(report_date) AS max_report_date
FROM read_parquet(
    'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/**/*.parquet'
)
WHERE month = '2026-03'
""").fetchdf()

print("Query 2 — March 2026 window")
march_window


# Query 3 — GSC availability using IS TRUE
gsc_availability = con.sql("""
SELECT
    COUNT(*) AS available_rows
FROM read_parquet(
    'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/**/*.parquet'
)
WHERE month = '2026-03'
  AND gsc_data_available IS TRUE
""").fetchdf()

print("Query 3 — GSC availability")
gsc_availability


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Query 1 — Grain
Query 2 — March 2026 window
Query 3 — GSC availability


,available_rows
0,3611061


### Deliberate leakage experiment

I deliberately add the future-derived April impressions to the feature set to demonstrate target leakage. April impressions are not available at the March decision moment, so this feature is not valid for a real prediction setting.

In [44]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

leak_df = decision_frame.dropna(
    subset=["april_impressions", "declining_proxy"]
).copy()

features = [
    "gsc_impressions",
    "gsc_clicks",
    "gsc_avg_position",
    "ga4_sessions",
    "ga4_engaged_sessions"
]

X_honest = leak_df[features]
X_leaky = leak_df[features + ["april_impressions"]]
y = leak_df["declining_proxy"]

X_train_h, X_test_h, y_train, y_test = train_test_split(
    X_honest,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

X_train_l, X_test_l, _, _ = train_test_split(
    X_leaky,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

honest_model = DecisionTreeClassifier(
    max_depth=4,
    random_state=42
)

honest_model.fit(X_train_h, y_train)
honest_pred = honest_model.predict(X_test_h)
honest_score = accuracy_score(y_test, honest_pred)

leaky_model = DecisionTreeClassifier(
    max_depth=4,
    random_state=42
)

leaky_model.fit(X_train_l, y_train)
leaky_pred = leaky_model.predict(X_test_l)
leaky_score = accuracy_score(y_test, leaky_pred)

print(f"Honest score: {honest_score:.3f}")
print(f"Leaky score:  {leaky_score:.3f}")

Honest score: 0.807
Leaky score:  0.824


### Leakage result

The honest model scored 0.807, while the model using the future-derived april_impressions feature scored 0.824.

The higher score is not evidence of a better real-world model. april_impressions is future information that would not be known at the March decision moment. This demonstrates target leakage.

The future-derived feature is removed from the final feature set, and the honest score is retained as the decision-support result.

## 4. Data limits
This slice has several limits. The history is unbalanced because the number of rows grows substantially over time. Early months also have less complete GSC coverage, so availability is not uniform across the panel. The future-outcome proxy depends on the April window, so it cannot be used as a feature at the March decision moment. In addition, overlapping time windows can make observations related rather than fully independent.

The results should therefore be treated as directional decision-support evidence, not as a causal estimate or a guarantee of future performance.

In [45]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
print("March decision rows:", len(decision_frame))
print(
    "Rows with April outcome available:",
    decision_frame["april_impressions"].notna().sum()
)


March decision rows: 331437
Rows with April outcome available: 331436


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.